### ISSUE TITLE
Maintaining the same context between decodeRequest and encodeResponse

### ISSUE THREAD

**Issue by HelloWorldMisc** [go-kit/kit#1131]
Hi,

When the request is being decoded at the transport layer, we are creating the context and adding some values in it. We are populating this context down to the service and below layers to extract some non param values from it. Now, when the response is received back at the endpoint layer, the context is preserved. However, when we are trying to encodeResponse() at the transport layer, the context values are no longer present. So, is there a way to preserve the context till the encoding of response is done? 
To summarize, this is the flow. DecodeRequest->create ctx & add values->pass the ctx to endpoint->pass the ctx to business logic layers... -> and the same ctx -> encodeResponse

Thanks for your time.

**Comment by sagikazarmark** [comment]
From your description it's not immediately obvious where exactly you "create" the context, but I'm guessing you do it in the endpoint layer.

Normally the context is not mutable, meaning that when you add something to the context it actually creates a new context and the parent context remains untouched. This is actually quite a good design.

One thing you could do is wrapping the context before decoding the request using a server before hook. In that hook you could inject a map in the context or rather use a library like this: https://github.com/peterbourgon/ctxdata

The bottom line is: attach a mutable structure to the context in a server before hook. You can add your data in there later and you should still be able to access it when you encode the request.

Here is the relevant code, it should be fairly obvious how it works: https://github.com/go-kit/kit/blob/a6c5d5805b360802a934766a63c254c9b94114f9/transport/http/server.go#L110-L132

Hope that helps!

**Comment by peterbourgon** [comment]
Yeah, that :)
